gui/seedwatch.lua

```diff
@@ -0,0 +1,310 @@
+
+local gui = require('gui')
+local widgets = require('gui.widgets')
+local plugin = require('plugins.seedwatch')
+
+local PROPERTIES_HEADER = '   Quantity     Target   '
+local REFRESH_MS = 10000
+local MAX_TARGET = 2147483647
+--
+-- SeedSettings
+--
+SeedSettings = defclass(SeedSettings, widgets.Window)
+SeedSettings.ATTRS{
+    lockable=false,
+    frame={l=5, t=5, w=35, h=9},
+    data={id='', name='', quantity=0, target=0,},
+}
+
+function SeedSettings:init()
+    self:addviews{
+        widgets.Label{
+            frame={t=0, l=0},
+            text='Seed: ',
+        },
+        widgets.Label{
+            view_id='id',
+            frame={t=0, l=0},
+            visible=false,
+        },
+        widgets.Label{
+            view_id='name',
+            frame={t=0, l=6},
+            text_pen=COLOR_GREEN,
+        },
+        widgets.Label{
+            frame={t=1, l=0},
+            text='Quantity: ',
+        },
+        widgets.Label{
+            view_id='quantity',
+            frame={t=1, l=10},
+            text_pen=COLOR_GREEN,
+        },
+        widgets.EditField{
+            view_id='target',
+            frame={t=2, l=0},
+            label_text='Target: ',
+            key='CUSTOM_CTRL_T',
+            on_char=function(ch) return ch:match('%d') end,
+            on_submit=self:callback('commit'),
+        },
+        widgets.HotkeyLabel{
+            frame={t=4, l=0},
+            key='SELECT',
+            label='Apply',
+            on_activate=self:callback('commit'),
+        },
+    }
+end
+
+function SeedSettings:show(choice, on_commit)
+    self.data = choice.data
+    self.on_commit = on_commit
+    local data = self.data
+    self.subviews.id:setText(data.id)
+    self.subviews.name:setText(data.name)
+    self.subviews.quantity:setText(tostring(data.quantity))
+    self.subviews.target:setText(tostring(data.target))
+    self.visible = true
+    self:setFocus(true)
+    self:updateLayout()
+end
+
+function SeedSettings:hide()
+    self:setFocus(false)
+    self.visible = false
+end
+
+function SeedSettings:commit()
+    local target = math.tointeger(self.subviews.target.text)
+    if not target or target == '' then
+        target = 0
+    elseif target > MAX_TARGET then
+        target = MAX_TARGET
+    end
+    plugin.seedwatch_setTarget(self.subviews.id.text, target)
+    self:hide()
+    self.on_commit()
+end
+
+function SeedSettings:onInput(keys)
+    if keys.LEAVESCREEN or keys._MOUSE_R_DOWN then
+        self:hide()
+        return true
+    end
+    SeedSettings.super.onInput(self, keys)
+    return true
+end
+
+--
+-- Seedwatch
+--
+Seedwatch = defclass(Seedwatch, widgets.Window)
+Seedwatch.ATTRS {
+    frame_title='Seedwatch',
+    frame={w=60, h=27},
+    resizable=true,
+    resize_min={h=25},
+    hide_unmonitored=DEFAULT_NIL,
+    manual_hide_unmonitored_touched=DEFAULT_NIL,
+    data = {sum=0, seeds_qty=0, seeds_watched=0, seeds = {{id='', name='', quantity=0, target=0,}}},
```

Comment:
internal data that is not intended to be set by a parent should be initialized in `init()`

Reply:
removed.